scrapy_poet/overrides.py

```diff
@@ -89,35 +70,64 @@ class OverridesRegistry(OverridesRegistryBase):
-                raise ValueError(
-                    f"Invalid rule: {rule}. Rules as tuples must have "
-                    f"3 elements: (1) the pattern, (2) the PO class used as a "
-                    f"replacement and (3) the PO class to be replaced."
-                )
-            pattern, use, instead_of = rule
-            rule = ApplyRule(
-                for_patterns=Patterns([pattern]), use=use, instead_of=instead_of
-            )
-        self.rules.append(rule)
-        # FIXME: This key will change with the new rule.to_return
-        self.matcher[rule.instead_of].add_or_update(  # type: ignore
-            len(self.rules) - 1, rule.for_patterns
+    def __init__(self, rules: Optional[Iterable[ApplyRule]] = None) -> None:
+        super().__init__(rules=rules)
+        self.overrides_matcher: Dict[Type[ItemPage], URLMatcher] = defaultdict(
+            URLMatcher
         )
+        self.item_matcher: Dict[Optional[Type], URLMatcher] = defaultdict(URLMatcher)
+        for rule_id, rule in enumerate(self._rules):
+            self.add_rule(rule_id, rule)
+        logger.debug(f"List of parsed ApplyRules:\n{self._rules}")
+
+    def add_rule(self, rule_id: int, rule: ApplyRule) -> None:
+        # A common case when a page object subclasses another one with the same
+        # URL pattern. See the ``test_item_return_subclass()`` test case in
+        # ``tests/test_web_poet_rules.py``.
+        matched = self.item_matcher[rule.to_return]
+        pattern_dupes = [
+            pattern
+            for pattern in matched.patterns.values()
+            if pattern == rule.for_patterns
+        ]
+        if pattern_dupes:
+            rules = [
+                r
+                for p in pattern_dupes
+                for r in self.search(for_patterns=p, to_return=rule.to_return)
+            ]
+            warn(
+                f"Similar URL patterns {pattern_dupes} were declared earlier "
+                f"that use to_return={rule.to_return}. The first, highest-priority "
+                f"rule added to SCRAPY_POET_REGISTRY will be used when matching "
+                f"against URLs. Consider updating the priority of these rules: "
+                f"{rules}."
+            )
 
-    def overrides_for(self, request: Request) -> Mapping[Callable, Callable]:
-        overrides: Dict[Callable, Callable] = {}
-        for instead_of, matcher in self.matcher.items():
-            rule_id = matcher.match(request.url)
+        if rule.instead_of:
+            self.overrides_matcher[rule.instead_of].add_or_update(
+                rule_id, rule.for_patterns
+            )
+        if rule.to_return:
+            self.item_matcher[rule.to_return].add_or_update(rule_id, rule.for_patterns)
+
+    # TODO: These URL matching functionalities could be moved to web-poet.
+    # ``overrides_for`` in web-poet could be ``str`` or ``_Url`` subclass.
+
+    def rules_for_url(
+        self, url: str, url_matcher: Dict[Any, URLMatcher]
```

Comment:
url_matcher -> matchers?

Reply:
Fixed in https://github.com/scrapinghub/web-poet/pull/112